numpy/core/src/multiarray/textreading/str_to_int.c

```diff
@@ -23,7 +23,7 @@ const char *deprecation_msg = (
 
 #define DECLARE_TO_INT(intw, INT_MIN, INT_MAX, byteswap_unaligned)          \
     NPY_NO_EXPORT int                                                       \
-    to_##intw(PyArray_Descr *descr,                                         \
+    npy_to_##intw(PyArray_Descr *descr,                                         \
```

Comment:
Not that it matters, but it is misaligned now ;).

In any case, these changes don't hurt and I will probably just put them in in a bit unless anyone beats me to it or disagrees.

(I have to admit, I am surprised by a statically linked toolchain, I mean in principle our public API is exported as an API table, of course the symbols in it match the static ones, but...)

Reply:
Yes, indeed things work when built that way with some fairly small changes. The unintended consequence is that C symbols NumPy thinks are module-private may end up exported and cause ODR problems. This rarely causes problems (but sometimes does, e.g., here). But I figured the fix makes the code no worse and arguably better.

(I would understand if you rejected this change on the grounds that it's only a problem with a non-standard build of NumPy. But I figured there's no harm asking.

I don't know if we've written much externally about how Google builds Python binaries, but the brief summary is that, somewhat like PyOxidizer, for every Python binary we deploy we build a copy of the Python interpreter and all C extension modules linked into a single binary. This gives us a hermetic deployment artifact and allows for things like cross-module link-time optimization.)

Reply:
I agree with @seberg, this seems harmless enough.

I'm curious, have you also fixed the google code to avoid generic names such as `tokenize` and `to_float`?  Is there now a `google_tokenize` instead of `tokenize` somewhere in the google codebase?

Reply:
It's actually typically not our code that has this issue, instead it's typically a problem with collisions between different parts of the Python extension ecosystem. As I mentioned, these are surprisingly rare, but they do happen! I noticed these particular collisions when upgrading to NumPy 1.23 a while back, but I don't recall a collision for many previous versions (since at least 1.16 or earlier).

One main reason for this is we write far more C++ than C, and this is typically only a problem for C code not C++ code, because C++ code has namespaces making collisions much less likely: you'd have to pick matching namespaces as well.

For example,

* `tokenize` appears in `astropy`: https://github.com/astropy/astropy/blob/df3e5bb36120797065c8e326309663ab1a3df788/astropy/io/ascii/src/tokenizer.h#L104
* `to_double` appears in `pandas`: https://github.com/pandas-dev/pandas/blob/ad190575aa75962d2d0eade2de81a5fe5a2e285b/pandas/_libs/src/parse_helper.h#L16

I could rename either of the colliding symbols (or carry a patch locally, which is what I'm doing right now). But obviously I prefer to upstream any changes rather than carry a patch if they make sense; this one falls on the boundary of what may or may not make sense upstream.

Reply:
Got it, thanks @hawkinsp.